### ISSUE TITLE
Docker version 20.10 breaks Harbor's connection to LDAP server

### ISSUE THREAD

**Issue by entasisj** [goharbor/harbor#13718]
**Expected behavior and actual behavior:**
Harbor should successfully connect to configured LDAP server.

**Steps to reproduce the problem:**
Update Docker CE package version to latest: 20.10 (which was released on **2020-12-08**).

**Versions:**
- harbor versions: 1.10.6, 2.1.1
- docker engine version: 20.10
- docker-compose version: 1.25.3 (with harbor 1.10.6), 1.27.4 (with harbor 2.1.1)

**Additional context:**
Log in 'core' when testing LDAP connection manually via Button in UI:
> Dec  9 11:19:21 172.22.0.1 core[8330]: 2020-12-09T10:19:21Z [WARNING] [/core/auth/ldap/ldap.go:58]: ldap connection fail: LDAP Result Code 200 "": dial tcp: i/o timeout
Dec  9 11:19:21 172.22.0.1 core[8330]: 2020-12-09T10:19:21Z [ERROR] [/core/controllers/base.go:109]: Error occurred in UserLogin: LDAP Result Code 200 "": dial tcp: i/o timeout
Dec  9 11:19:27 172.22.0.1 core[8330]: 2020-12-09T10:19:27Z [ERROR] [/lib/http/error.go:54]: {"errors":[{"code":"INTERNAL_SERVER_ERROR","message":"LDAP connect fail, error: LDAP Result Code 200 \"\": dial tcp: i/o timeout"}]}

**Comment by nicoherbigde** [comment]
We have the same problem as @entasisj that we can't resolve our internal LDAP server and mail server via DNS.
Via IP the access to both systems works without problems.

**Version:**

- harbor versions: 2.1.1
- docker engine version: 20.10
- docker-compose version: 1.26.0

Our current workaround is that for each Docker Service in the docker-compose.yml` we have specified our DNS servers in addition to the Docker internal DNS.

```
core:
  image: goharbor/harbor-core:v2.1.1
  container_name: harbor-core
  ...
  dns:
    - 127.0.0.11
    - <IP address of internal company DNS server>
  ...
```

Perhaps there is a more elegant interim solution.

**Comment by sluetze** [comment]
Docker changed their DNS behavior:
https://github.com/moby/moby/pull/41022

This may or may not cause this errors. can you say something about your OS and if systemd-resolvd is used or something else?

**Comment by nicoherbigde** [comment]
Hi @sluetze, we run our Docker setup on Ubuntu 18.04 LTS with the official APT package source: `https://download.docker.com/linux/ubuntu bionic stable`.

If I start a single Docker container via:

`docker run -it --rm --entrypoint /bin/bash --user root goharbor/harbor-core:v2.1.1`

and take a look inside the `/etc/resolv.conf` the following output is shown:

```
nameserver <IP address of our first internal company DNS server>
nameserver <IP address of our second internal company DNS server>
```

These two internal company DNS servers are also the ones we have registered inside the docker host. Furthermore, if I install the package `iputils` inside this docker container and then ping our LDAP and mail server, they will be resolved correctly.

On the other hand, when I log into our Harbor installation via:

`docker-compose exec -u root core bash`

and take a look inside the `/etc/resolv.conf` the following output is shown:

```
nameserver 127.0.0.11
options edns0 ndots:0
```

When I try to install the package `iputils`, I already get the following error messages.

```
tdnf install iputils

Error: Failed to synchronize cache for repo 'VMware Photon Extras 2.0(x86_64)' from 'https://packages.vmware.com/photon/2.0/photon_extras_2.0_x86_64'
Disabling Repo: 'VMware Photon Extras 2.0(x86_64)'
Error: Failed to synchronize cache for repo 'VMware Photon Linux 2.0(x86_64)' from 'https://packages.vmware.com/photon/2.0/photon_release_2.0_x86_64'
Disabling Repo: 'VMware Photon Linux 2.0(x86_64)'
Error: Failed to synchronize cache for repo 'VMware Photon Linux 2.0(x86_64) Updates' from 'https://packages.vmware.com/photon/2.0/photon_updates_2.0_x86_64'
Disabling Repo: 'VMware Photon Linux 2.0(x86_64) Updates'
No package iputils available
Error(1011) : No matching packages
```

In sum, resolving the DNS names does not work.

**Comment by rselph** [comment]
I believe a related behavior is causing trouble for my configuration as well.  Perhaps it provides some insight here.

Let's say I'm running docker on host foo.localdomain.  There are containers on that host that reference this address to connect to services.  But because we are now using the systemd resolver, foo.localdomain now resolves to 127.0.0.1 since the system resolver treats its own host name as loopback. This obviously doesn't work from inside the container.  To my mind, this is a breaking change.  I could perhaps add an alternative host name to my main DNS server, or connect directly to the host's IP address instead.  But neither of these solutions is very satisfying.  Running as network=host would work as well, but that breaks encapsulation.

**Comment by stonezdj** [comment]
You could solve this issue by define dns in the docker host machine, /etc/docker/daemon.json
```
{
 "dns":["<your dns server"]
}
```
and restart docker daemon.
```
sudo systemd restart docker
```
